crates/chia-sdk-driver/src/primitives/cat.rs

```diff
@@ -1,171 +1,152 @@
+        let cat_layer = CatLayer::new(self.asset_id, inner_spend.puzzle);
 
-        let solution_ptr = thing.construct_solution(
+        let puzzle_ptr = cat_layer.construct_puzzle(ctx)?;
+        let solution_ptr = cat_layer.construct_solution(
             ctx,
             CatSolution {
-                lineage_proof,
+                lineage_proof: self.lineage_proof,
                 prev_coin_id,
                 this_coin_info,
                 next_coin_proof,
                 prev_subtotal,
                 extra_delta,
-                inner_puzzle_solution: inner_spend.solution(),
+                inner_puzzle_solution: inner_spend.solution,
             },
         )?;
-        let solution =
-            Program::from_node_ptr(ctx.allocator(), solution_ptr).map_err(DriverError::FromClvm)?;
 
-        let cs = CoinSpend {
-            coin: self.coin,
-            puzzle_reveal: puzzle,
-            solution,
+        let puzzle = ctx.serialize(&puzzle_ptr)?;
+        let solution = ctx.serialize(&solution_ptr)?;
+
+        Ok(CoinSpend::new(self.coin, puzzle, solution))
+    }
+
+    /// Returns the lineage proof that would be used by each child.
+    pub fn child_lineage_proof(&self) -> LineageProof {
+        LineageProof {
+            parent_parent_coin_info: self.coin.parent_coin_info,
+            parent_inner_puzzle_hash: self.p2_puzzle_hash,
+            parent_amount: self.coin.amount,
+        }
+    }
+
+    /// Creates a wrapped spendable CAT for a given output.
+    #[must_use]
+    pub fn wrapped_child(&self, p2_puzzle_hash: Bytes32, amount: u64) -> Self {
+        let puzzle_hash = CatArgs::curry_tree_hash(self.asset_id, p2_puzzle_hash.into());
+        Self {
+            coin: Coin::new(self.coin.coin_id(), puzzle_hash.into(), amount),
+            lineage_proof: Some(self.child_lineage_proof()),
+            asset_id: self.asset_id,
+            p2_puzzle_hash,
+        }
+    }
+}
+
+impl Primitive for Cat {
+    fn from_parent_spend(
+        allocator: &mut Allocator,
+        parent_coin: Coin,
+        parent_puzzle: Puzzle,
+        parent_solution: NodePtr,
+        coin: Coin,
+    ) -> Result<Option<Self>, DriverError>
+    where
+        Self: Sized,
+    {
+        let Some(parent_layer) = CatLayer::<Puzzle>::parse_puzzle(allocator, parent_puzzle)? else {
+            return Ok(None);
+        };
+        let parent_solution = CatLayer::<Puzzle>::parse_solution(allocator, parent_solution)?;
+
+        let output = run_puzzle(
+            allocator,
+            parent_layer.inner_puzzle.ptr(),
+            parent_solution.inner_puzzle_solution,
+        )?;
+        let conditions = Vec::<Condition>::from_clvm(allocator, output)?;
+
+        let p2_puzzle_hash = conditions
+            .into_iter()
+            .filter_map(Condition::into_create_coin)
+            .find_map(|create_coin| {
+                // This is an optimization to skip calculating the hash.
```

Comment:
you're 5 steps ahead at all times

Reply:
Might be a premature optimization but I figured it can't hurt 😂